alibi/explainers/similarity/grad.py

```diff
@@ -0,0 +1,203 @@
+"""Gradient-based explainer.
+
+This module implements the gradient-based explainers grad-dot and grad-cos.
+"""
+
+import copy
+from typing import TYPE_CHECKING, Callable, Optional, Union, Dict, Tuple
+
+import numpy as np
+
+from alibi.api.interfaces import Explanation
+from alibi.explainers.similarity.base import BaseSimilarityExplainer
+from alibi.explainers.similarity.metrics import dot, cos, asym_dot
+from alibi.api.defaults import DEFAULT_META_SIM, DEFAULT_DATA_SIM
+
+if TYPE_CHECKING:
+    import tensorflow
+    import torch
+
+
+class SimilarityExplainer(BaseSimilarityExplainer):
+    def __init__(self,
+                 predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
+                 loss_fn: '''Callable[[Union[tensorflow.Tensor, torch.Tensor],
+                                    Union[tensorflow.Tensor, torch.Tensor]],
+                                   Union[tensorflow.Tensor, torch.Tensor]]''',
+                 sim_fn: 'Union[Callable[[np.ndarray, np.ndarray], np.ndarray], str]' = 'grad_dot',
+                 task: str = "classification",
+                 store_grads: bool = False,
+                 seed: int = 0,
+                 backend: str = "tensorflow",
+                 **kwargs
+                 ):
+        """Constructor
+
+        Parameters
+        ----------
+        predictor:
+            model to explain.
+        loss_fn:
+            Loss function used.
+        sim_fn:
+            Similarity function to use. ``'grad_dot'`` | ``'grad_cos'``. Default: ``'grad_dot'``.
+        task:
+            Task to perform. ``'classification'`` | ``'regression'``. Default: ``'classification'``.
+        store_grads:
+            Whether to store gradients. Default ``False``. If ``False``, gradients are computed on the fly otherwise we
+            store them which can be faster when it comes to explaining and instances.
```

Comment:
What are the drawbacks of storing gradients? Also "... explaining and instances." doesn't feel correct?